presto-hive/src/main/java/com/facebook/presto/hive/S3SelectCsvRecordReader.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.amazonaws.services.s3.AmazonS3URI;
+import com.amazonaws.services.s3.model.CSVInput;
+import com.amazonaws.services.s3.model.CSVOutput;
+import com.amazonaws.services.s3.model.CompressionType;
+import com.amazonaws.services.s3.model.ExpressionType;
+import com.amazonaws.services.s3.model.InputSerialization;
+import com.amazonaws.services.s3.model.OutputSerialization;
+import com.amazonaws.services.s3.model.SelectObjectContentRequest;
+import com.facebook.presto.spi.predicate.TupleDomain;
+import com.facebook.presto.spi.type.TypeManager;
+import org.apache.hadoop.fs.Path;
+import parquet.hadoop.metadata.CompressionCodecName;
+
+import java.util.List;
+import java.util.Properties;
+
+import static org.apache.hadoop.hive.serde.serdeConstants.ESCAPE_CHAR;
+import static org.apache.hadoop.hive.serde.serdeConstants.QUOTE_CHAR;
+
+public class S3SelectCsvRecordReader
+        extends S3SelectLineRecordReader
+{
+    // Sentinel unicode comment character (http://www.unicode.org/faq/private_use.html#nonchar_codes)
+    private static final String COMMENTS_CHAR_STR = "\uFDD0";
```

Comment:
What's the role of this? Is the value of `COMMENTS_CHAR_STR` important? eg. can it be "\uFDD1" or "a" ?

Reply:
S3Select allows customers to specify [row level comment character](http://lahiri.aka.corp.amazon.com/bzip2_builddev/selecting-content-from-objects.html) for their csv files. Default value for the character is "#".

For the S3Optimizer, we use hive metastore or AWS Glue to fetch schema information for the object. As per our understanding, the schema object doesn't have support for row level comment character (https://github.com/ogrodnek/csv-serde). 

Currently, there is no clean way of specifying through S3Select API that we want to disable CSV comment characters, thus a sentinel noncharacter unicode value is used to act as a proxy to turn off row level comment identification.

## Example:

### Input
#### Presto query
```
select * from test_data;
```

#### Transformed IonSql query
```
select * from S3Object s
```
* S3Object s refers to the S3 object pointed by test_data table in hive metastore or AWS Glue.

### CSV Object Data 
```
1,2,3
a,b,c
#This is a comment, ignore this, used comment character(#)
x,y,z
```

### Expected Output for Presto with S3 Get
```
1,2,3
a,b,c
#This is a comment, ignore this, used comment character(#)
x,y,z
```

### Actual Output for Presto with S3Select with "#" as comment character string
```
1,2,3
a,b,c
x,y,z
```

* If we use "a" as the comment character string then it would result in following output for the above query and dataset
```
1,2,3
#This is a comment, ignore this, used comment character(#)
x,y,z
```

Both of the above results are incorrect.

Reply:
To answer your question, it could be any character that is a reserved noncharacter in unicode spec (http://www.unicode.org/faq/private_use.html#nonchar_codes). This is to ensure that customers don't use it intentionally in their files. Another requirement is that it could be encoded with UTF-8 to construct S3Select XML request string as described in https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectSELECTContent.html

Reply:
Please add a comment that it's expected that this "sentinel character" isn't part of the actual users' data. 
Also, add a TODO comment, since it should not be assumed that users' data doesn't contain \uFDD0 (even though this is a "noncharacter", this still is a valid code point and legal to be used or stored in a data file).

Reply:
Sounds good. How about the following:

```
// It is expected that \uFDD0 sentinel comment character is not the first character in any row of user's CSV S3 object. The rows starting with \uFDD0 will be skipped by S3Select and will not be a part of the result set or aggregations.
// To process CSV objects that may contain \uFDD0 as first row character please disable S3Optimizer.
```

What is the TODO here?

Reply:
Your comment is good. The TODO -- i meant something like `// TODO remove this when S3Select adds support for declaring that there is no character introducing a comment`